File: Programmers/yerim/다리를 지나는 트럭.cpp
// [42583] 다리를 지나는 트럭
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int ans = 0;
    int weightSum = 0; // 다리 위의 트럭 무게의 합
    queue<int> q;
    
    for(int i = 0; i < truck_weights.size(); i++) {
        while(true) {
            if(q.size() == bridge_length) { // 트럭 개수 최대
                weightSum -= q.front(); // 맨 앞의 트럭 제거
                q.pop();
            }
            if(truck_weights[i] + weightSum <= weight) {
                break;
            }
            q.push(0); // 무게가 L을 넘는 경우
            ans++;
        }
        q.push(truck_weights[i]);
        weightSum += truck_weights[i];
        ans++;
    }
    return ans + bridge_length;
}

File: Programmers/bumjin/다리를지나는트럭.java
package Programmers.bumjin;

import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;

        //모든 트럭이 건너려면 최소 몇 초

        //대기트럭
        Queue<Truck> wait = new LinkedList();
        //지나가는 트럭
        List<Truck> list = new ArrayList();

        for (int i : truck_weights) {
            wait.add(new Truck(i, bridge_length));
        }

        int currentWeight = 0;
        int possibleWeight = weight;

        while (!(wait.isEmpty() && list.isEmpty())) {
            if (!list.isEmpty() && list.get(0).distance <= 0) {
                //다리 위에서 제거
                possibleWeight += list.get(0).weight;//다리 무게 원상복귀
                list.remove(0);
            }

            if (!wait.isEmpty() && possibleWeight - wait.peek().weight >= 0) {
                possibleWeight -= wait.peek().weight;
                list.add(wait.poll());
            }

            for (int i=0; i<list.size(); i++) {
                Truck truck = list.get(i);
                truck.distance--;
            }
            answer++;
        }

        return answer;
    }

    static class Truck {
        int weight, distance;

        Truck(int weight, int distance) {
            this.weight = weight;
            this.distance = distance;
        }
    }
}
